File: bytopic/hashtable/java/group1/RansomNote.java
// Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the
// letters from magazine and false otherwise.
// Each letter in magazine can only be used once in ransomNote.
// Example 1:
// Input: ransomNote = "aa", magazine = "aab"
// Output: true
// Example 2:
// Input: ransomNote = "aa", magazine = "ab"
// Output: false
// Tag: 39/150

class RansomNote {
    public static void main(String[] args) {
        String ransomNote = "aa";
        String magazine = "aab";
        System.out.println("Can construct? " + canConstruct(ransomNote, magazine));
    }

    static boolean canConstruct(String ransomNote, String magazine) {
        if (ransomNote.length() > magazine.length()) return false;
        int[] alphabetsCounter = new int[26];

        for (char c: magazine.toCharArray())
            alphabetsCounter[c - 'a']++;

        for (char c: ransomNote.toCharArray()) {
            if (alphabetsCounter[c - 'a'] == 0) return false;
            alphabetsCounter[c - 'a']--;
        }

        return true;
    }
}

File: bytopic/hashtable/python/group1/ransom_note.py
"""
Given two strings ransomNote and magazine, return true if ransomNote can be constructed by using the
letters from magazine and false otherwise.
Each letter in magazine can only be used once in ransomNote.

Example 1:
Input: ransomNote = "aa", magazine = "aab"
Output: true
Example 2:
Input: ransomNote = "aa", magazine = "ab"
Output: false

Tag: 39/150
Tag: 383/2927, R502/2936 (overall frequency ranking)
"""
from collections import Counter


def can_construct(ransom_note: str, magazine: str) -> bool:
    st1, st2 = Counter(ransom_note), Counter(magazine)
    if st1 & st2 == st1:
        return True
    return False


def main():
    ransom_note = "aa"
    magazine = "ab"
    print(can_construct(ransom_note, magazine))


if __name__ == "__main__":
    main()
